### ISSUE TITLE
Feature request: Add FreeBSD to build targets

### ISSUE THREAD

**Issue by PaddyMac** [adventuregamestudio/ags#1265]
The current version of AGS for FreeBSD is 3.4.4.1 from 2019. I'd like to have a newer version available. I've been tinkering with it, and after lying to the build system and making it think it's building on Linux, everything works fine 0n 3.5.0.31 if I use GCC up until the final linking phase when I get some undefined reference errors. There will need to be a little extra work to get it working with Clang which is the FreeBSD native compiler. In any case, I'm not much of a programmer, but I'd certainly be willing to put in some effort on my part to get this working. Somewhat tangentially, could you make the build system accept the use of more system libraries? Other than Allegro, it doesn't look like you're using customized versions of external libraries. I mention this partly because I tried compiling the 3.6 series of AGS, and I ran into issues compiling the included SDL2 library which I also ran into previously when trying to manually compile an unpatched SDL2 on FreeBSD. So this issue could be avoided entirely if it were possible to use the system library instead of a bundled one.

**Comment by rofl0r** [comment]
if you use Engine/Makefile, only strictly required library components are built - to my knowledge that's only the embedded freetype copy. i've been using that makefile since forever with system-provided allegro4 library, there was never a need for a especially patched version. you should however use allegro 4.4.3 as it has all a number of bugfixes and features to support AGS better.
as for SDL2 version aka 3.6: the code uses libsdl-sound in the SDL2 incarnation, which has so far not seen a release (however one was recently promised to debian maintainers in a mail that has been linked somewhere here). you gotta build that on your own from mercurial/git and install it systemwide, but apart from that you can use Engine/Makefile just fine.

> Add FreeBSD to build targets

i don't think a freebsd-specific build target needs to be added, it should build just fine using the makefile, unless there are some `#ifdef __linux__` or similar defines, in which case maybe the "linux target" should be remodeled as a "generic posix" target.

i guess a good start would be if you report the specific errors you're experiencing so those can be examinated on a case-by-case basis.

**Comment by ivan-mogilko** [comment]
@PaddyMac 
I'd like to clarify, you are refering to Cmake scripts and want to add FreeBSD to our Cmake scripts, is that correct?

As far as I know it's Cmake that fetches our patched version of Allegro 4 for Linux, even though today this may not be completely necessary (at least from the point of functionality). AGS should work well with their official latest release (4.4.3.*) which already contains all the patches we were sending to them over last couple several years.
But engine should also link with the older 4.4.2 version of Allegro 4 as well, except in this case couple of functions will not work as expected (noteably FLIC playback, and maybe something else I forgot about).
Strictly speaking, our patched Allegro 4 is only required for Windows version of AGS because of couple of hacks, but that is not relevant to this topic.

In regards to 3.6.0, guess there's similar problem that cmake script fetches particular SDL2 revision and wants to build it itself.
We already had a suggestion for an option to use system libraries instead: https://github.com/adventuregamestudio/ags/issues/1148#issuecomment-817320639

Perhaps we could add similar option for Allegro 4 in 3.5.* branch?

**Comment by ericoporto** [comment]
> unless there are some #ifdef __linux__

I think they're on both alconfig.h in allegro (maybe only on the SDL2 port), and in the Common/core/platform.h . I don't remember if there's more at this moment.

TBH I think it's better to add FreeBSD as a separate target instead of going the generic posix route - or create a Unix target that for now only has FreeBSD and in the future add more to it.

**Comment by PaddyMac** [comment]
[ags-cmake-fail.txt](https://github.com/adventuregamestudio/ags/files/6470042/ags-cmake-fail.txt)
I attached the console output so show what I got, but the configure phase failed. And the failures are, as far as I can tell, SDL-related. The most recent version of SDL2 in the FreeBSD repo is 2.0.12, and I know it has a few patches applied and some features disabled to get it working right. 2.0.14 is in the pipeline to get in the repo. Just to see, I tried the git head version of SDL2, and it configured with some non-fatal warnings and compiled just fine. I don't know what version of SDL is included in the AGS source tree, but presumably if it's 2.0.14 or earlier, it may need to have some patches applied to compile under FreeBSD. At the very least, it might need some additional cmake parameters to disable unsupported features. I'm not sure why you didn't hit any errors. I still think using the system SDL would help avoid these errors, but maybe using the head version in the AGS tree would work too.

**Comment by ericoporto** [comment]
Yes, it's the right branch. No there should be no need for any additional parameters or patches. Because I have two days of experience with FreeBSD, I need some more information - the error stack points to a CMake file in your system, not in ags or sdl2.

Which version of FreeBSD are you using? I didn't knew if there was a right one so I got the 13 from here (https://www.freebsd.org/releases/13.0R/announce/). And which CMake version are you using? Have you modified it in any way? Can you remove and `pkg install cmake` again?

**Comment by ivan-mogilko** [comment]
>  I don't know what version of SDL is included in the AGS source tree

To clarify, SDL is not included in the AGS source tree. It's fetched by cmake scripts, but that's optional.

**Comment by ivan-mogilko** [comment]
Regarding the error, I am not an expert in CMake, but simply by error message it looks like not liking the format of the command line arguments passed in particular check.
E.g. I found similar error case discussed here: https://stackoverflow.com/questions/57114900/how-to-pass-an-argument-with-no-value-to-cmake

Could this be a difference in CMake version and/or mode?

**Comment by ivan-mogilko** [comment]
@ericoporto @PaddyMac the error in the log looks like CMake complaining on some command syntax. Not a Cmake expert, but I did a quick search and found this recent commit in the sdl2 repository: https://github.com/libsdl-org/SDL/commit/34de3b57a101f8d885041add5e21311a153bb1bf

If I understand correctly, they replace "-D_THREAD_SAFE" to "-D_THREAD_SAFE=1" format.
This replacement is called in all the GLES-related checks.

**Comment by ericoporto** [comment]
The CMake module `CheckCSourceCompiles.cmake` doesn't exist in my FreeBSD install, this is why I asked the version.

**Comment by morganwillcock** [comment]
I think in terms of platform checks within the application they will likely have to be OS specific, that is each BSD system would need to be identified separately since they aren't compatible in the same way that Linux distributions are (there is already a working OpenBSD port for AGS too but from memory I think they apply patches to the Linux specific parts and pretend to be Linux). There are also a few surrounding issues outside the engine, like the launch wrapper assuming the use of GNU readlink.

In terms of build system I think it is probably better to try and be more generic, that includes building the programs, but also all the surrounding interactions that are currently spread between lots of CI containers, and generating a better source release so that anyone wanting to try packaging it has an easier time. Given the Debian packaging is tangled with the existing build processes and current CI issues I'm probably talking myself into giving it a try.

**Comment by PaddyMac** [comment]
@ericoporto I tried your build. At startup, I got a message saying "Unable to load theora video 's7lo800.ogv'", but after hitting a key the game proceeded to load and play. I only tested for a couple of minutes, but it seemed fine. I think the theora error simply kept a short intro video from playing. I'll attach a log of the console output which has some debugging messages. 
[catladyags.txt](https://github.com/adventuregamestudio/ags/files/6476254/catladyags.txt)
The cmake version I have is 3.20.1; I have not modified it. I removed and reinstalled the package, but that made no difference. The FreeBSD version I'm using is 13.0-RELEASE which is the newest stable version.

You can take a look at https://github.com/freebsd/freebsd-ports/tree/main/devel/sdl20 in the files/ directory to see how SDL 2.0.12 (which is what AGS's cmake script fetches) is patched for FreeBSD as well as the Makefile to see how it's configured. Maybe that will help shed light on the issues that the unpatched 2.0.12 release has on FreeBSD and how to address them.

**Comment by ericoporto** [comment]
This is super weird because theoretically we have the same version of things. I will try to figure out on the weekend. But it really should build fine.

The debug messages come from SDL2, we haven't disabled those.

The game you are running apparently can't find a plugin that is direct3d related but I think it's alright. About the video I need to check. Is it Cat Lady Steam version?

The ports build is not on our end and I have mailed the appropriate people about it. This will take some time.

**Comment by ivan-mogilko** [comment]
@ericoporto have your PR #1324 resolve this, or there's something else necessary?

**Comment by ericoporto** [comment]
I think the issue meant a real package in FreeBSD ports, but this requires a stable release (non-alpha and non-beta), and someone updating the port there. I would say the necessary "plumbing" is done by the PR, so closing this issue should be fine.

**Comment by ivan-mogilko** [comment]
@PaddyMac I apologize for bringing this up after so long time, but were you able to resolve the issue on your side?
I was not able to follow all the conversation here, as my knowledge on the topic is quite small.
Which version are you currently using on FreeBSD? Are there any more issues on our side that require fixing?

**Comment by ericoporto** [comment]
https://cgit.freebsd.org/ports/commit/?id=05203b94b144224c53d048556a9ba255f2fc20d6

Just to note here in case someone finds this issue on Google, it looks like FreeBSD AGS package now has a maintainer

https://cgit.freebsd.org/ports/commit/?id=05203b94b144224c53d048556a9ba255f2fc20d6
